### PR TITLE
Single root element only

### PR DESCRIPTION
**Pull request by NorbertHuethmayr**
# Purpose / Goal
This PR fixes issue #207 

# Type
Please mention the type of PR
* [x] Bug Fix
* [ ] Refactoring / Technology upgrade
* [ ] New Feature

### CODE REVIEW

**Comment by amitguptagwl** [comment]
Will this code work for repeated nodes? 
```xml
<root></root><root></root>
```

**Comment by NorbertHuethmayr** [comment]
> Will this code work for repeated nodes?
> 
> ```
> <root></root><root></root>
> ```

Hi Amit,

the code you mentioned is exactly what is invalid as it does not have a single root element.
If multiple nested elements are contained within a single root element, then yes, it will work.

I also added 2 unit tests to verify just that.

Best regards
Norbert

**Comment by amitguptagwl** [comment]
Your PR is testing
```xml
<xml></xml><xml2></xml2>
```
and
```xml
<test>
        <nested>
        </nested>
</test>
<xml2>
</xml2>
```
but not
```xml
<xml></xml><xml></xml>
```

Please also read my comment on the issue you raised for the same.

**Comment by NorbertHuethmayr** [comment]
You are right - the code i have pushed is not working properly. I will work on a better fix....

**Comment by NorbertHuethmayr** [comment]
@amitguptagwl 

I've improved the detection so that the depth of a node is tracked. If the root level is reached more than once, an error (including the line) is created.
I also had to change your "large" file as it was technically invalid (it had multiple root nodes)

Let me know what you think of this PR.

Best regards
Norbert

File: src/validator.js [hunk 1]
```diff
@@ -131,10 +135,19 @@ exports.validate = function (xmlData, options) {
...
+            if(reachedRoot)
+            {
+              return getErrorObject('InvalidXml', 'Multiple possible root nodes found.', getLineNumberForPosition(xmlData, i));
+            }
 
-          //we remember the first opening tag that we find
-          if (firstOpeningTag === '') {
-            firstOpeningTag = tagName;
+            //otherwise we set a flag for the next time around
+            reachedRoot = true;
```
Comment (amitguptagwl) [inline src/validator.js]:
```js
if (tags.length === 0) {
    reachedRoot = true;
}
```

File: src/validator.js [hunk 2]
```diff
@@ -131,10 +135,19 @@ exports.validate = function (xmlData, options) {
           }
           tags.push(tagName);
```
Comment (amitguptagwl) [inline src/validator.js]: if(reachedRoot === true){
    return getErrorObject('InvalidXml', 'Multiple possible root nodes found.', getLineNumberForPosition(xmlData, i));
}else{
    tags.push(tagName);
}

Comment (NorbertHuethmayr) [reply src/validator.js]: Good point, will do...

**Comment by amitguptagwl** [comment]
Changes look fine in the validator. I'll check other files nad merge the changes as soon as possible.

**Comment by NorbertHuethmayr** [comment]
> Big file increases the of the repo and user faces the issue at the time of cloning. Hence, I was wondering if you can remove the new large file. And use some code like this in the test;
> 
> ```js
> const xmlData = fs.readFileSync(fileNamePath).toString();
> //xmlData=`<root>${xmlData.repeat(1000)}</root>`;
> ```
> 
> In this way, we can even control the size of the data on the fly. Eventually, we'll remove other big files as well from the project.

Hi Amit,

I pretty much disagree with you there. The size of a repo cant be a concern. Imagine really big projects with thousands of files....

One thing that I could do is to cut out half of the content from the copied file.
Would that be ok?

**Comment by amitguptagwl** [comment]
> Hi Amit,
>
>I pretty much disagree with you there. The size of a repo cant be a concern. Imagine really big projects with thousands of files....
>
>One thing that I could do is to cut out half of the content from the copied file.
Would that be ok?

A big project is different than a project with bigger files. Large XMl that we have in our project is of 100mb. You must be having good internet speed, I don't. I sometimes need to clone the project of all the PRs to test them thoroughly which requires more time and bandwidth.

In addition, this is not that something we should compromise due to any limitations. It gives us more control.

File: spec/validator_spec.js [hunk 3]
```diff
@@ -511,4 +511,33 @@ attribute2="attribute2"
...
+        </test>
+        <xml2>
+        </xml2>`;
+
+        const expected = { code: "InvalidXml", msg: 'Multiple possible root nodes found.', line: 5};
+        var result = validator.validate(xmlData).err;
+        expect(result).toEqual(expected);
+    });
+
+    it('should not validate XML documents with multiple root nodes 3', () => {
```
[the remark this answers is not in the document]
Comment (NorbertHuethmayr) [reply spec/validator_spec.js]: Done

**Comment by NorbertHuethmayr** [comment]
> ```js
> when root nodes have more nested tags
> ```

Hi Amit,

I introduced a new group and grouped the existing tests as you suggested.

**Comment by NorbertHuethmayr** [comment]
> > Hi Amit,
> > I pretty much disagree with you there. The size of a repo cant be a concern. Imagine really big projects with thousands of files....
> > One thing that I could do is to cut out half of the content from the copied file.
> > Would that be ok?
> 
> A big project is different than a project with bigger files. Large XMl that we have in our project is of 100mb. You must be having good internet speed, but all the contributors might not have. I sometimes need to clone the project of all the PRs to test them thoroughly which requires more time and bandwidth.
> 
> In addition, this is not that something we should compromise due to any limitations. It gives us more control. Think it in this way, what is the advantage of having large files in a project when you can eliminate them?
> 
> Or think it to optimize the existing available assets. Eg. you can use `ptest.xml` and make it as bigger as you want in the test as per your requirement. I would be happy to hear the cons of this approach.

I've removed the 2nd very large file and the unit test that used it. It was a duplicate test anyway and didnt do much in the first place.

**Comment by amitguptagwl** [comment]
Thanks for your contribution.

**Pull request merged**